File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
private:
    bool isPalindrome(int low , string &s , int high){
        while(low < high){
             if(s[low++] != s[high--]) return false;
        }
        return true;
    }
  void findPartition(int index,string s , vector<vector<string>> &result,vector<string> container){
        if(index == s.size()){
            result.push_back(container);
            return;
        }

        for(int i=index;i<s.size(); i++){
            if(isPalindrome(index,s,i)){
                container.push_back(s.substr(index,i-index+1));
                findPartition(i+1,s,result,container);
                container.pop_back();
            }
        }
  }
public:
    vector<vector<string>> partition(string s) {
       vector<vector<string>> result;
       vector<string> container;
       findPartition(0,s,result,container);
       return result;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    private boolean isPalindrome(int start, int end, String s){
        char[] ch = s.toCharArray();
        while(start <= end){
            if(ch[start] != ch[end]){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
    private void findPartition(int index, List<List<String>> result, String s,List<String> temp){
        if(index == s.length()){
            result.add(new ArrayList<>(temp));
            return;
        }

        for(int i=index; i<s.length(); i++){
            if(isPalindrome(index,i,s)){
                String curr = s.substring(index,i+1);
                temp.add(curr);
                findPartition(i+1,result,s,temp);
                temp.remove(temp.size()-1);
            }
        }
    }
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        List<String> temp = new ArrayList<>();
        findPartition(0,result,s,temp);
        return result;
    }
}